File: Editorial/0142_Linked_List_Cycle_II/C/2_Floyd_s_Tortoise_and_Hare_Algorithm.c
struct ListNode *detectCycle(struct ListNode *head) {
    // Initialize tortoise and hare pointers
    struct ListNode *tortoise = head;
    struct ListNode *hare = head;
    if (head == NULL || head->next == NULL) {
        return NULL;
    }
    // Move tortoise one step and hare two steps
    do {
        tortoise = tortoise->next;
        hare = hare->next->next;
    } while (hare != NULL && hare->next != NULL && tortoise != hare);
    // Check if there is no cycle
    if (hare == NULL || hare->next == NULL) {
        return NULL;
    }
    // Reset the hare pointer to the head
    hare = head;
    // Move both pointers one step until they meet again
    while (tortoise != hare) {
        tortoise = tortoise->next;
        hare = hare->next;
    }
    // Return the node where the cycle begins
    return tortoise;
}

File: Editorial/142_Linked_List_Cycle_II/Go/2_Floyd_s_Tortoise_and_Hare_Algorithm.go
func detectCycle(head *ListNode) *ListNode {
    // Initialize tortoise and hare pointers
    tortoise, hare := head, head
    // Move tortoise one step and hare two steps
    for hare != nil && hare.Next != nil {
        tortoise = tortoise.Next
        hare = hare.Next.Next
        // Check if the hare meets the tortoise
        if tortoise == hare {
            break
        }
    }
    // Check if there is no cycle
    if hare == nil || hare.Next == nil {
        return nil
    }
    // Reset either the hare pointer to the head
    hare = head
    // Move both pointers one step until they meet again
    for tortoise != hare {
        tortoise = tortoise.Next
        hare = hare.Next
    }
    // Return the node where the cycle begins
    return tortoise
}
